### PR TITLE
CDAP-4571 Multiple system artifact directories.

### PR DESCRIPTION
**Pull request by anwar6953**
Support for system plugins being picked up from multiple directories.

https://issues.cask.co/browse/CDAP-4571
http://builds.cask.co/browse/CDAP-RBT575-2

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java [hunk 1]
```diff
@@ -493,31 +496,33 @@ public void addSystemArtifacts() throws IOException, WriteConflictException {
...
-      // parse id from filename
-      Id.Artifact artifactId;
-      try {
-        artifactId = Id.Artifact.parse(Id.Namespace.SYSTEM, jarFile.getName());
-      } catch (IllegalArgumentException e) {
-        LOG.warn(String.format("Skipping system artifact '%s' because the name is invalid: ", e.getMessage()));
-        continue;
-      }
+    for (File systemArtifactDir : systemArtifactDirs) {
+      for (File jarFile : DirUtils.listFiles(systemArtifactDir, "jar")) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java]: we weren't doing this before, but lets add some sanity checks, make sure the dir exists and is a directory. If not, log a warning that it is invalid and will be ignored.

**Comment by albertshau** [comment]
minor comment about adding some error handling. Otherwise lgtm

**Comment by anwar6953** [comment]
Updated

**Comment by albertshau** [comment]
minor comment about the error message. lgtm

**Comment by gokulavasan** [comment]
@anwar6953 Can you add some info about the why this change is necessary? JIRA would be even better.

**Comment by anwar6953** [comment]
Filed JIRA, and put the use case in the description. In summary, users may have plugins installed in multiple locations, and this allows them avoid moving/copying the JARs to a central location.
https://issues.cask.co/browse/CDAP-4571

**Comment by gokulavasan** [comment]
Thanks @anwar6953. I thought there was some other deeper use case driving this. Thanks for filing a JIRA!

**Comment by anwar6953** [comment]
Squashed commits. Pending build.

**Comment by bdmogal** [comment]
@anwar6953 are there any other configuration params that take delimited string-list values? Do we have a standard separator for them? I mean do we always use a semi-colon as the delimiter?

**Comment by anwar6953** [comment]
`kafka.seed.brokers` is comma-separated.
```
<property>
    <name>kafka.seed.brokers</name>
    <value>127.0.0.1:9092</value>
    <description>
      List of Kafka brokers (comma separated)
    </description>
  </property>
```

I chose to use semicolon as a separator for the directory paths because comma can be in a valid file path.

**Comment by anwar6953** [comment]
@bdmogal Any concerns with that? The alternative I see is to use colon as the separator.
In PATH, semicolon and colon are standard delimiters for Windows and Linux, respectively.

**Comment by bdmogal** [comment]
``yarn.application.classpath`` uses comma as the delimiter - https://hadoop.apache.org/docs/r2.4.1/hadoop-yarn/hadoop-yarn-common/yarn-default.xml. 

I'm ok, as long as we use something standard across all delimited path valued configuration params

**Comment by anwar6953** [comment]
comma can be in the name of a file (I just checked on linux CentOS).
This is the first delimited path value configuration, so I think its fine as long as we are consistent with new ones in the future.

**Comment by bdmogal** [comment]
:+1:

**Pull request merged**